### PR TITLE
feat(map): codegen map type

### PR DESCRIPTION
**Pull request by Jonathan-Casey**
### C#
```
   public Dictionary<string, string> dictionary1 { get; set; };
   public Dictionary<string, System.DateTime> dictionary2 { get; set; };
   public Dictionary<string, Employee> dictionary3 { get; set; };
```

### Typescript
```
   export type Map1 = Map<string, string>;
   export type Map2 = Map<string, Date>;
   export type Map3 = Map<string, IEmployee>;
```

Note: Key Type is generated as Type String across all supported Targets until ambiguity on `$class` property is resolved with consensus amongst core concerto maintainers.

### CODE REVIEW

**Comment by mttrbrts** [comment]
> Key Type is generated as Type String across all supported Targets until ambiguity on $class property is resolved with consensus amongst core concerto maintainers.

Which issue does this refer to?

File: test/codegen/fromcto/typescript/typescriptvisitor.js [hunk 1]
```diff
@@ -629,6 +641,100 @@ describe('TypescriptVisitor', function () {
...
+            let param = {
+                fileWriter: mockFileWriter
+            };
+
+            let mockMapDeclaration = sinon.createStubInstance(MapDeclaration);
+
+            const getKeyType = sinon.stub();
+            const getValueType = sinon.stub();
+
+            getKeyType.returns('String');
```
Comment (ekarademir) [inline test/codegen/fromcto/typescript/typescriptvisitor.js:725]: Should this return `DateTime`?

Comment (Jonathan-Casey) [reply test/codegen/fromcto/typescript/typescriptvisitor.js:725]: @ekarademir see **Note** in the preview description.

File: test/common/graph.js [hunk 2]
```diff
@@ -106,7 +110,10 @@ describe('graph', function () {
             const connectedGraph = graph.findConnectedGraph('org.acme.hr@1.0.0.ChangeOfAddress');
             expect(connectedGraph.hasEdge('org.acme.hr@1.0.0.ChangeOfAddress', 'org.acme.hr@1.0.0.Person'));
 
-            const filteredModelManager = modelManager.filter(declaration => connectedGraph.hasVertex(declaration.getFullyQualifiedName()));
+
+            const filteredModelManager = modelManager
+                .filter(declaration => !declaration.isMapDeclaration?.())
```
[the remark this answers is not in the document]
Comment (Jonathan-Casey) [reply test/common/graph.js:115]: Given this is "just" a test case, my preference is on readability vs overloading the predicate.

File: test/codegen/fromcto/csharp/csharpvisitor.js [hunk 3]
```diff
@@ -1485,6 +1500,68 @@ public class SampleModel : Concept {
...
+            findStub.returns(mockMapDeclaration);
+            getKeyType.returns('String');
+            getValueType.returns('DateTime');
+            mockField.getName.returns('Map1');
+            mockMapDeclaration.getName.returns('Map1');
+            mockMapDeclaration.isMapDeclaration.returns(true);
+            mockMapDeclaration.getKey.returns({ getType: getKeyType });
+            mockMapDeclaration.getValue.returns({ getType: getValueType });
+            csharpVisitor.visitField(mockField, param);
+            param.fileWriter.writeLine.withArgs(1, 'public Dictionary<string, string> Map1 { get; set; };').calledOnce.should.be.ok;
```
[the remark this answers is not in the document]
Comment (Jonathan-Casey) [reply test/codegen/fromcto/csharp/csharpvisitor.js]: Yes, it does now, you are reviewing a draft :-)

File: lib/codegen/fromcto/odata/odatavisitor.js [hunk 4]
```diff
@@ -140,11 +142,13 @@ class ODataVisitor {
...
         parameters.fileWriter.writeLine(1, `<EntityContainer Name="${namespace}Service">`);
-        modelFile.getAllDeclarations().forEach((decl) => {
-            if (!decl.isAbstract() && decl.isIdentified()) {
-                parameters.fileWriter.writeLine(2, `<EntitySet Name="${decl.getName()}" EntityType="${ModelUtil.removeNamespaceVersionFromFullyQualifiedName(decl.getFullyQualifiedName())}"/>`);
-            }
-        });
+        modelFile.getAllDeclarations()
+            .filter(decl => !decl.isMapDeclaration?.())
+            .forEach((decl) => {
+                if (!decl.isAbstract() && decl.isIdentified()) {
```
[the remark this answers is not in the document]
Comment (Jonathan-Casey) [reply lib/codegen/fromcto/odata/odatavisitor.js:148]: See comment above on test case. I'd argue this hinders readability, and offers no true performance benefit given the size of the data footprint we typically process in any given ModelFile.

[the remark this answers is not in the document]
Comment (mttrbrts) [reply lib/codegen/fromcto/odata/odatavisitor.js:148]: This one is a niggle for me too. Personally, I find it more legible to do all of the filtering together.

**Review by mttrbrts — APPROVED** [review]

**Pull request merged**